// @metamask/snaps-cli/dist/cmds/build/bundle.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundle = void 0;
const browserify_1 = __importDefault(require("browserify"));
const builders_1 = require("../../builders");
const bundleUtils_1 = require("./bundleUtils");
// We need to statically import all Browserify transforms and all Babel presets
// and plugins, and calling `require` is the sanest way to do that.
/* eslint-disable @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires, node/global-require */
/**
 * Builds a Snap bundle JSON file from its JavaScript source.
 *
 * @param src - The source file path.
 * @param dest - The destination file path.
 * @param argv - arguments as an object generated by yargs.
 * @param argv.sourceMaps - Whether to output sourcemaps.
 * @param argv.stripComments - Whether to remove comments from code.
 * @param argv.transpilationMode - The Babel transpilation mode.
 */
function bundle(src, dest, argv) {
    const { sourceMaps: debug, transpilationMode } = argv;
    return new Promise((resolve, _reject) => {
        const bundleStream = bundleUtils_1.createBundleStream(dest);
        const bundler = browserify_1.default(src, { debug });
        if (transpilationMode !== builders_1.TranspilationModes.none) {
            bundler.transform(require('babelify'), {
                global: transpilationMode === builders_1.TranspilationModes.all,
                presets: [
                    [
                        require('@babel/preset-env'),
                        {
                            targets: {
                                browsers: ['chrome >= 66', 'firefox >= 68'],
                            },
                        },
                    ],
                ],
                plugins: [
                    require('@babel/plugin-transform-runtime'),
                    require('@babel/plugin-proposal-class-properties'),
                    require('@babel/plugin-proposal-object-rest-spread'),
                    require('@babel/plugin-proposal-optional-chaining'),
                    require('@babel/plugin-proposal-nullish-coalescing-operator'),
                ],
            });
        }
        bundler.bundle(async (bundleError, bundleBuffer) => await bundleUtils_1.closeBundleStream({
            bundleError,
            bundleBuffer,
            bundleStream,
            src,
            dest,
            resolve,
            argv,
        }));
    });
}
exports.bundle = bundle;
