// @metamask/snap-utils/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandlerType = exports.SNAP_STREAM_NAMES = exports.SnapValidationFailureReason = exports.NpmSnapFileNames = exports.SnapIdPrefixes = void 0;
/**
 * The possible prefixes for snap ids.
 */
var SnapIdPrefixes;
(function (SnapIdPrefixes) {
    SnapIdPrefixes["npm"] = "npm:";
    SnapIdPrefixes["local"] = "local:";
})(SnapIdPrefixes = exports.SnapIdPrefixes || (exports.SnapIdPrefixes = {}));
var NpmSnapFileNames;
(function (NpmSnapFileNames) {
    NpmSnapFileNames["PackageJson"] = "package.json";
    NpmSnapFileNames["Manifest"] = "snap.manifest.json";
})(NpmSnapFileNames = exports.NpmSnapFileNames || (exports.NpmSnapFileNames = {}));
/**
 * Snap validation failure reason codes that are programmatically fixable
 * if validation occurs during development.
 */
var SnapValidationFailureReason;
(function (SnapValidationFailureReason) {
    SnapValidationFailureReason["NameMismatch"] = "\"name\" field mismatch";
    SnapValidationFailureReason["VersionMismatch"] = "\"version\" field mismatch";
    SnapValidationFailureReason["RepositoryMismatch"] = "\"repository\" field mismatch";
    SnapValidationFailureReason["ShasumMismatch"] = "\"shasum\" field mismatch";
})(SnapValidationFailureReason = exports.SnapValidationFailureReason || (exports.SnapValidationFailureReason = {}));
var SNAP_STREAM_NAMES;
(function (SNAP_STREAM_NAMES) {
    SNAP_STREAM_NAMES["JSON_RPC"] = "jsonRpc";
    SNAP_STREAM_NAMES["COMMAND"] = "command";
})(SNAP_STREAM_NAMES = exports.SNAP_STREAM_NAMES || (exports.SNAP_STREAM_NAMES = {}));
var HandlerType;
(function (HandlerType) {
    HandlerType["OnRpcRequest"] = "onRpcRequest";
    HandlerType["OnTransaction"] = "onTransaction";
})(HandlerType = exports.HandlerType || (exports.HandlerType = {}));
