// @metamask/snaps-cli/dist/cmds/serve/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const http_1 = __importDefault(require("http"));
const serve_handler_1 = __importDefault(require("serve-handler"));
const builders_1 = __importDefault(require("../../builders"));
const utils_1 = require("../../utils");
const serveUtils_1 = require("./serveUtils");
/**
 * Starts a local, static HTTP server on the given port with the given root
 * directory.
 *
 * @param argv - arguments as an object generated by yargs
 * @param argv.root - The root directory path string
 * @param argv.port - The server port
 */
async function serve(argv) {
    const { port, root: rootDir } = argv;
    await (0, utils_1.validateDirPath)(rootDir, true);
    console.log(`\nStarting server...`);
    const server = http_1.default.createServer(async (req, res) => {
        await (0, serve_handler_1.default)(req, res, {
            public: rootDir,
            headers: [
                {
                    source: '**/*',
                    headers: [
                        {
                            key: 'Cache-Control',
                            value: 'no-cache',
                        },
                    ],
                },
            ],
        });
    });
    server.listen({ port }, () => (0, serveUtils_1.logServerListening)(port));
    server.on('request', (request) => (0, serveUtils_1.logRequest)(request));
    server.on('error', (error) => {
        (0, serveUtils_1.logServerError)(error, argv.port);
        process.exitCode = 1;
    });
    server.on('close', () => {
        console.log('Server closed');
        process.exitCode = 1;
    });
}
module.exports = {
    command: ['serve', 's'],
    desc: 'Locally serve Snap file(s) for testing',
    builder: (yarg) => {
        yarg.option('root', builders_1.default.root).option('port', builders_1.default.port);
    },
    handler: (argv) => serve(argv),
};
