// @metamask/snaps-cli/dist/cmds/init/initUtils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.yarnInstall = exports.gitInit = exports.isInGitRepository = exports.isGitInstalled = exports.cloneTemplate = exports.prepareWorkingDirectory = exports.SNAP_LOCATION = exports.TEMPLATE_GIT_URL = void 0;
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const path_1 = __importDefault(require("path"));
exports.TEMPLATE_GIT_URL = 'https://github.com/MetaMask/template-snap-monorepo.git';
exports.SNAP_LOCATION = 'packages/snap/';
/**
 * Checks if the destination folder exists and if it's empty. Otherwise create it.
 *
 * @param directory - The desination folder.
 */
async function prepareWorkingDirectory(directory) {
    try {
        const isCurrentDirectory = directory === process.cwd();
        if (!isCurrentDirectory) {
            try {
                await fs_1.promises.mkdir(directory, { recursive: true });
            }
            catch (error) {
                throw new Error('Init Error: Failed to create new directory.');
            }
        }
        const existingFiles = await fs_1.promises.readdir(directory);
        if (existingFiles.length > 0) {
            throw new Error(`Directory ${directory} not empty.`);
        }
    }
    catch (error) {
        throw new Error(`Init Error: Failed to prepare working directory with message: ${error.message}`);
    }
}
exports.prepareWorkingDirectory = prepareWorkingDirectory;
/**
 * Clones the template in a directory.
 *
 * @param directory - The directory to clone the template in.
 */
function cloneTemplate(directory) {
    try {
        (0, child_process_1.execSync)(`git clone --depth=1 ${exports.TEMPLATE_GIT_URL} ${directory}`, {
            stdio: [2],
        });
    }
    catch (error) {
        throw new Error('Init Error: Failed to clone the template.');
    }
}
exports.cloneTemplate = cloneTemplate;
/**
 * Check if git is installed.
 *
 * @returns True if git is installed, or false otherwise.
 */
function isGitInstalled() {
    try {
        (0, child_process_1.execSync)('git --version', { stdio: 'ignore' });
        return true;
    }
    catch (error) {
        return false;
    }
}
exports.isGitInstalled = isGitInstalled;
/**
 * Check if the actual working dir is a git repository.
 *
 * @param directory - The directory to check.
 * @returns True if it's a git repository otherwise false.
 */
function isInGitRepository(directory) {
    try {
        (0, child_process_1.execSync)('git rev-parse --is-inside-work-tree', {
            stdio: 'ignore',
            cwd: path_1.default.resolve(__dirname, directory),
        });
        return true;
    }
    catch (error) {
        return false;
    }
}
exports.isInGitRepository = isInGitRepository;
/**
 * Init a git repository.
 *
 * @param directory - The directory to init.
 */
function gitInit(directory) {
    try {
        (0, child_process_1.execSync)('git init', {
            stdio: 'ignore',
            cwd: path_1.default.resolve(__dirname, directory),
        });
    }
    catch (error) {
        throw new Error('Init Error: Failed to init a new git repository.');
    }
}
exports.gitInit = gitInit;
/**
 * Install dependencies in a yarn project.
 *
 * @param directory - The directory containing the project.
 */
function yarnInstall(directory) {
    try {
        (0, child_process_1.execSync)('yarn install', {
            stdio: [0, 1, 2],
            cwd: path_1.default.resolve(__dirname, directory),
        });
    }
    catch (error) {
        throw new Error('Init Error: Failed to install dependencies.');
    }
}
exports.yarnInstall = yarnInstall;
